File: api/src/resolvers/money.ts
import { Resolvers } from "../resolvers-types"

export const Money: Resolvers["Money"] = {
  formatted: (money) => {
    const absoluteWithCommas = money
      .toString()
      .replace(/^-/, "")
      .replace(/\B(?=(\d{3})+(?!\d))/g, ",")

    return `${money.amount < 0 ? "-" : ""}${money.getCurrencyInfo().symbol}${absoluteWithCommas}`
  },

  formattedShort: (money) => {
    const absoluteAmount = Math.round(Math.abs(money.toDecimal()))
    let amount = absoluteAmount
    let suffix = ""

    if (absoluteAmount > 1_000_000) {
      amount = Math.round(absoluteAmount / 1_000_000)
      suffix = "M"
    } else if (absoluteAmount > 10_000) {
      amount = Math.round(absoluteAmount / 1_000)
      suffix = "K"
    }

    return `${money.amount < 0 ? "-" : ""}${money.getCurrencyInfo().symbol}${amount}${suffix}`
  },

  integerAmount: (money) => Math.round(money.amount),
  decimalAmount: (money) => money.toDecimal()
}

File: backend/app/concepts/money.rb
class Money < Data.define(:amount_cents, :currency)
  def +(other)
    raise ArgumentError, "Currency mismatch" unless currency == other.currency

    self.class.new(**self.to_h, amount_cents: amount_cents + other.amount_cents)
  end

  def -(other)
    raise ArgumentError, "Currency mismatch" unless currency == other.currency

    self.class.new(**self.to_h, amount_cents: amount_cents - other.amount_cents)
  end

  def abs
    self.class.new(**self.to_h, amount_cents: amount_cents.abs)
  end

  def amount_decimal
    amount_cents.to_d / (10 ** currency.decimal_digits)
  end

  def formatted
    "#{sign}#{currency.symbol}#{format('%.2f', amount_decimal.abs)}"
  end

  def formatted_short
    amount = amount_decimal.abs.round
    suffix = ""

    if amount > 1_000_000
      amount = (amount / 1_000_000).round
      suffix = "M"
    elsif amount > 10_000
      amount = (amount / 1_000).round
      suffix = "K"
    end

    return "#{sign}#{currency.symbol}#{amount}#{suffix}"
  end

  def on_date(date)
    MoneyOnDate.new(currency:, date:, amount_cents:)
  end

  private

  def sign
    amount_cents < 0 ? "-" : ""
  end
end
